libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/LocationUpdater.java

```diff
@@ -0,0 +1,87 @@
+package com.mapbox.services.android.navigation.v5.navigation;
+
+import android.annotation.SuppressLint;
+import android.location.Location;
+import android.support.annotation.NonNull;
+
+import com.mapbox.android.core.location.LocationEngine;
+import com.mapbox.android.core.location.LocationEngineCallback;
+import com.mapbox.android.core.location.LocationEngineRequest;
+import com.mapbox.android.core.location.LocationEngineResult;
+
+import java.lang.ref.WeakReference;
+
+import timber.log.Timber;
+
+class LocationUpdater {
+
+  private final LocationEngineCallback<LocationEngineResult> callback = new CurrentLocationEngineCallback(this);
+  private final RouteProcessorBackgroundThread thread;
+  private LocationEngine locationEngine;
+  private LocationEngineRequest request;
+
+  @SuppressLint("MissingPermission")
+  LocationUpdater(RouteProcessorBackgroundThread thread, LocationEngine locationEngine,
+                  LocationEngineRequest request) {
+    this.thread = thread;
+    this.locationEngine = locationEngine;
+    this.request = request;
+    requestInitialLocationUpdates(locationEngine, request);
+  }
+
+  void updateLocationEngine(LocationEngine locationEngine) {
+    requestLocationUpdates(request, locationEngine);
+    this.locationEngine = locationEngine;
+  }
+
+  void updateLocationEngineRequest(LocationEngineRequest request) {
+    requestLocationUpdates(request, locationEngine);
+    this.request = request;
+  }
+
+  void onLocationChanged(Location location) {
+    if (location != null) {
+      thread.updateLocation(location);
+      NavigationTelemetry.getInstance().updateLocation(location);
+    }
+  }
+
+  void removeLocationUpdates() {
+    locationEngine.removeLocationUpdates(callback);
+  }
+
+  @SuppressLint("MissingPermission")
+  private void requestInitialLocationUpdates(LocationEngine locationEngine, LocationEngineRequest request) {
+    locationEngine.getLastLocation(callback);
```

Comment:
What happens in the situation in which the last location is really far from where you actually are? Could this end up in a "undesired" flying effect? I'm wondering if we should / could remove it 🤔